tests/utils/print_vec_mat.hpp

```diff
@@ -207,5 +213,174 @@ void print_matrix( const grb::Matrix< T, B > & mat, size_t _limit = 0, const cha
+ * @param mat matrix to print
+ * @param _limit max number of rows and columns to print (0 for all)
+ * @param head optional heading to print \b before the matrix
+ */
+template<
+	typename T,
+	enum grb::Backend B,
+	typename std::enable_if< std::is_void< T >::value, int >::type = 0
+>
+void print_matrix( const grb::Matrix< T, B > & mat, size_t _limit = 0, const char * head = nullptr ) {
+	const size_t rows = grb::nrows( mat );
+	const size_t cols = grb::ncols( mat );
+	size_t row_limit = _limit == 0 ? rows : std::min( _limit, rows );
+	size_t col_limit = _limit == 0 ? cols : std::min( _limit, cols );
+	// create and dump only relevant portion
+	dense_mat< bool > assigned( row_limit, col_limit );
+	for( const auto& t : mat ) {
+		auto row = t.first;
+		auto col = t.second;
+		assigned[ row ][ col ] = ( row < row_limit && col < col_limit );
+	}
+
+	if( head != nullptr ) {
+		std::cout << "<<< " << head << " >>>" << std::endl;
+	}
+	std::cout << "=== PATTERN-MATRIX ===" << std::endl;
+	std::cout << "Size: " << rows << " x " << cols << std::endl;
+	for( size_t i = 0; i < row_limit; ++i ) {
+		for( size_t j = 0; j < col_limit; ++j ) {
+			bool has_value = assigned[ i ][ j ];
+			if( has_value ) {
+				std::cout << "X";
+			} else {
+				std::cout << "_";
+			}
+			std::cout << " ";
+		}
+		std::cout << std::endl;
+	}
+	std::cout << "==============" << std::endl << std::endl;
+}
+
+namespace
+{
+	template<
+		typename D,
+		class Storage,
+		typename std::enable_if< std::is_void< D >::value, int >::type = 0
+	>
+	void printCompressedStorage(
+		const Storage & storage,
+		size_t n,
+		size_t nnz,
+		std::ostream & os = std::cout
+	) {
+		os << "  col_start (" << n + 1 << "): [ ";
+		for( size_t i = 0; i <= n; ++i ) {
+			os << storage.col_start[ i ] << " ";
+		}
+		os << "]" << std::endl;
+		os << "  row_index (" << nnz << "): \n[\n";
+		for( size_t i = 0; i < n; ++i ) {
+			os << " " << std::setfill( '0' ) << std::setw( 2 ) << i << ":  ";
+			for( auto t = storage.col_start[ i ]; t < storage.col_start[ i + 1 ]; t++ )
+				os << std::setfill( '0' ) << std::setw( 2 ) << storage.row_index[ t ] << " ";
+			os << std::endl;
+		}
+		os << "]" << std::endl;
+	}
+
+	template<
+		typename D,
+		class Storage,
+		typename std::enable_if< not std::is_void< D >::value, int >::type = 0
+	>
+	void printCompressedStorage(
+		const Storage & storage,
+		size_t n,
+		size_t nnz,
+		std::ostream & os
+	) {
+		printCompressedStorage< void >( storage, n, nnz, os );
+		os << "  values    (" << nnz << "): [ ";
+		for( size_t i = 0; i < nnz; ++i ) {
+			os << storage.values[ i ] << " ";
+		}
+		os << "]" << std::endl << std::flush;
+	}
+
+} // namespace
+
+
+template<
+	bool Enabled = true,
+	typename D, typename RIT, typename CIT, typename NIT,
+	Backend implementation,
+	typename = void
+>
+void printCRS(
+	const Matrix< D, implementation, RIT, CIT, NIT > &,
+	const std::string & = "",
+	std::ostream & = std::cout
+) {}
```

Comment:
is this dead code? (I probably don't totally get the pattern used here)

Same for the CCS variant below

Reply:
Got it=)